src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,392 @@
+//!     >>> clustering = AgglomerativeClustering().fit(X)
+//!     >>> clustering
+//!     AgglomerativeClustering()
+//!     >>> clustering.labels_
+//!     array([1, 1, 1, 0, 0, 0])
+//!
+//! ## Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::math::num::RealNumber;
+//! use smartcore::cluster::agglomerative::*;
+//! use smartcore::algorithm::neighbour::fastpair::FastPair;
+//!
+//! // Iris data
+//! let x = DenseMatrix::from_2d_array(&[
+//!            &[5.1, 3.5, 1.4, 0.2],
+//!            &[4.9, 3.0, 1.4, 0.2],
+//!            &[4.7, 3.2, 1.3, 0.2],
+//!            &[4.6, 3.1, 1.5, 0.2],
+//!            &[5.0, 3.6, 1.4, 0.2],
+//!            &[5.4, 3.9, 1.7, 0.4],
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! // Fit to data, with a threshold
+//! // example using FastPair, compute dendrogram
+//! let clustering = AggregativeFastPair::fit(&x).unwrap();
+//! let dendrogram = clustering.labels();
+//! ```
+//!
+use std::marker::PhantomData;
+
+use crate::algorithm::neighbour::fastpair::FastPair;
+use crate::error::Failed;
+use crate::linalg::Matrix;
+use crate::math::distance::euclidian::Euclidian;
+use crate::math::num::RealNumber;
+
+///
+/// Abstract trait for sequential, agglomerative, hierarchic, non-overlapping methods
+///
+pub trait SAHNClustering<T: RealNumber, M: Matrix<T>> {
+    ///
+    /// Aggregate the data according to given distance threshold
+    ///
+    fn fit(data: &M) -> Result<AggregativeFastPair<T, M>, Failed>;
+
+    ///
+    /// Return clusters labels
+    ///
+    fn labels(&self) -> &Vec<Vec<T>>;
+}
+
+///
+/// An implementation of Bottom-Up (Agglomerative) Hierarchical
+///  Clustering with `FastPair`
+///
+pub struct AggregativeFastPair<T: RealNumber, M: Matrix<T>> {
+    dendrogram: Vec<Vec<T>>, // computed labels
+    _marker: PhantomData<M>,
+}
+
+// Add linkage algorithms
+impl<T: RealNumber, M: Matrix<T>> FastCluster<T> for AggregativeFastPair<T, M> {}
+
+// Implement aggregative clustering using FastPair and fastcluster
+impl<T: RealNumber, M: Matrix<T>> SAHNClustering<T, M> for AggregativeFastPair<T, M> {
+    //
+    // 1. Compute `FastPair` on matrix's rows
+    // 2. Compute sparse dissimilarities into upper-trinagular matrix
+    // 3. Compute dendrogram
+    //
+    // The linkage distance threshold above which clusters will not be merged.
+    fn fit(data: &M) -> Result<AggregativeFastPair<T, M>, Failed> {
```

Comment:
Do you plan to add different types of linkage (“ward”, “complete”, “average”, “single”) later? Even if you plan to support only "ward" at this moment we still should have an enum parameter `linkage ` with a single item "ward"

Reply:
FastPair works only and only with single. To add different linkages we need also to add an alternative algorithm like NearestNeighbour that can work with all the linkages.

Reply:
I feel that it should be possible. Take a look at [this paper](https://arxiv.org/pdf/cs/9912014.pdf):

"We can perform median, centroid, Ward’s, or other
bottom-up clustering methods in which clusters are represented by objects, in time
O(n^2 log^2n) and space O(n)."

Reply:
I don't know if I understood wrong but I think that paper refers to fastcluster that is the clustering algorithm; FastPair is a closest neighbour algorithm. The correspective of FastPair in that paper is the Nearest Neighbour algorithm that can be used with other linkages.